rust/arrow/src/buffer.rs

```diff
@@ -511,12 +512,22 @@ impl MutableBuffer {
 
     /// Returns the data stored in this buffer as a slice.
     pub fn data(&self) -> &[u8] {
-        unsafe { std::slice::from_raw_parts(self.raw_data(), self.len()) }
+        if self.data.is_null() {
+            &[]
+        } else {
+            unsafe { std::slice::from_raw_parts(self.raw_data(), self.len()) }
+        }
     }
 
     /// Returns the data stored in this buffer as a mutable slice.
     pub fn data_mut(&mut self) -> &mut [u8] {
-        unsafe { std::slice::from_raw_parts_mut(self.raw_data() as *mut u8, self.len()) }
+        if self.data.is_null() {
+            &mut []
+        } else {
+            unsafe {
+                std::slice::from_raw_parts_mut(self.raw_data() as *mut u8, self.len())
+            }
+        }
     }
 
     /// Returns a raw pointer for this buffer.
```

Comment:
General comment about `raw_data` below; there should be a separate `&mut self` accessor for getting a `*mut u8` pointer. This is related to the rules of the stacked borrows model, and I'm not the best interpreter of those rules. Compare for example with `Vec::as_mut_ptr`. (Short explanation, this is part of the budding formalization of rules for raw pointers and `unsafe` code blocks.)